### ISSUE TITLE
Cheaters and trolls issues 

### ISSUE THREAD

**Issue by jasbrake** [ActionFPS/ActionFPS-Game#161]
This is an inexhaustive list of reasons AssaultCube has so many issues with cheating and trolling, and what I think could be done to fix this.

## Glitches
- **Reconnect spamming** - Usually used to avoid a ban, but sometimes used as a form of DOS attack. When combined with vote spamming, the perpetrator is usually untouchable unless an admin is present and very quick to ban them.
**Solution:** A cooldown period if they reconnect multiple times in a row and have the server keep track of the users vote cooldown, even if they disconnect. User registration will definitely help with solving this.

- **Vote Forgery** - This is when a user creates a vote with false information (a different teamkill count, a higher ping, etc). The only way I know how to do this is by inflating the vote reason with white space until the real information (ping, teamkills) get's pushed off the screen. I'm not sure if a more elaborate way has been developed, but I've seen versions with varying degrees of realism.
**Solution:** A more thorough trimming of the vote reason should prevent the method I know of. 

## Game Features

- **Lack of player registration** - the ability to instantly change to someone else and have practically no one be any the wiser is a huge encouragement towards abusive behavior. I think user registration is the first step towards encouraging user behavior and punishing abuse.
**Solution:** From what I understand, this is already developed and is currently in testing.

- **Lack of anticheat** - I know that some servers have a form of anticheat, but from what I understand it's minimal. The fact that admins cannot monitor their servers 24/7 means that AC needs some form of anticheat. I know that developing an anticheat is complex and it can usually always be beaten anyway. However, the obvious cheats still need to be caught. In my opinion, the easiest things to produce anticheat for and that every server should have implemented if it's going to connect to the MS are:
  - **Map glitching** - leaving the map boundaries, going through walls and under the floors. I believe some servers already have this.
  - **Flying** - I believe this is already implemented in some servers as well. Should be just checking to see if a player doesn't touch the ground for a certain amount of time.
  - **Teleporting/Speedhacking** - Again, already somewhat implemented (or maybe that's specifically a "Got the flag too fast" detection). But it needs to be implemented on all servers and be more aggressive. I understand it's hard to find a balance between not banning a user because they lag and banning teleporting users, but perhaps a combination of checks could be effective _enough_. For example: are they currently shooting, did they just take damage from a grenade, did they move too many cubes than should be possible without recoil/grenade explosions?
  - **Rapid Fire** - When the fire rate is increased. This can potentially be prevented by checking if a player sends too many packets in a certain timeframe.

- **Dev enforced builds** The easiest way to develop a hack for AC is simply to modify the source and recompile. 
**Solution:** Only allow builds signed by the developers to connect to servers.

- **Infrequent updates** This is a big reason why we have such a problem with cheaters and trolls. The game has some serious flaws, and nothing has been  done about them for years.

- **Lack of admin tools** - Everything that can be done to make admin's jobs easier, should be done (taking into account level of effort). For example: I use a script to see all the teamkills at once. This is something that should come with AC by default.
**Solution:** I want to implement some sort of server stats/management from the web for myself. It would be nice if I didn't have to login to a server to ban someone with 15 teamkills. Ideally anyone can add their own servers and manage their own. It would be nice if the servers supported something like this by default. Maybe have some sort of admin API served on a third port (the first 2 being the game itself and the server information).

## The nature of AssaultCube

- **It's a small codebase** - This makes reversing it and developing cheats easier because it takes less effort to find things. Naturally, larger codebases and games are harder to develop cheats for because they are usually more complex.
**Solution:** Add more features! Just kidding. This isn't something that has to change.

- **It's open source** - Although it's fairly obvious why this can be detrimental to stopping cheaters, this is not something I wish to change.
**Solution:** None required.

Feel free to suggest more issues and potential solutions.

**Comment by gyunderscorebe** [comment]
Another issue: when a vote is called to ban someone, they can simply disconnect and wait for the vote to fail, then reconnect. As a solution, ban votes should always pass when required number of F1s is reached, even if the player is disconnected.

**Comment by gyunderscorebe** [comment]
I think it would be a good idea to turn these into separate issues and then add checkboxes with links to them to this list.

**Comment by jasbrake** [comment]
Yes that's similar to reconnect spamming, except they just wait a little while to reconnect.

That's a good idea. I'd like to get some opinions on which of these issues should be fixed in the next release so I'm not just creating tickets willy-nilly.

**Comment by ScalaWilliam** [comment]
Game management problems will be alleviated with a programmable interface ("Admin interface"), either:
* A headless client or,
* Server modding using Lua, or both.

Player registration is already implemented, just the game is not released.

Before we take on anything further, the key blocker is [releasing the game](https://github.com/ActionFPS/ActionFPS-Game/milestone/2).

We don't need an anti-cheat in-client, it's a hopeless thing that we cannot fix unless we're funded. My solution to this would be to have an external analyser of behaviour to simply eliminate bad players using authentication.

Admin tools would be part of building an "admin interface". But this is blocked on the initial release.

The required work before we can proceed with improving in-game stuff is the first release. That's it :-)

**Comment by ScalaWilliam** [comment]
Today I merged the SDL2 branch. If you're available please test and file any bug reports.

:)
